### PR TITLE
Given pgCollision_LinePoint a "width"

### PR DESCRIPTION
**Pull request by ScriptLineStudios**
#14

### CODE REVIEW

**Comment by itzpr3d4t0r** [comment]
Can you run a line benchmark with and without this PR and send the resullts on discord?

**Comment by itzpr3d4t0r** [comment]
I'm a bit concerned that trading a ternary op. for three sqrts isn't the best choice

**Comment by ScriptLineStudios** [comment]
RESULTS:

COLLISION: LINE-LINE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 604.11  | 60.41  |   2.26    |
| Non colliding | 608.94  | 60.89  |   0.55    |

--------------------------------------------------
Group Mean: 60.65
Group Total: 1213.05
Group Standard Deviation: 0.24
Group Median: 60.65

COLLISION: LINE-CIRCLE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 589.19  | 58.92  |     1     |
| Non colliding | 776.01  |  77.6  |   1.11    |
| inside circle | 577.31  | 57.73  |   0.93    |

--------------------------------------------------
Group Mean: 64.75
Group Total: 1942.51
Group Standard Deviation: 9.1
Group Median: 58.92

COLLISION: LINE-LINE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 604.11  | 60.41  |   2.26    |
| Non colliding | 608.94  | 60.89  |   0.55    |

--------------------------------------------------
Group Mean: 60.65
Group Total: 1213.05
Group Standard Deviation: 0.24
Group Median: 60.65



COLLISION: LINE-CIRCLE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 589.19  | 58.92  |     1     |
| Non colliding | 776.01  |  77.6  |   1.11    |
| inside circle | 577.31  | 57.73  |   0.93    |

--------------------------------------------------
Group Mean: 64.75
Group Total: 1942.51
Group Standard Deviation: 9.1
Group Median: 58.92



COLLISION: LINE-POINT
==================================================
|         Name          |  Total  |  Mean  |  Std Dev  |
|-----------------------|---------|--------|-----------|
|    Colliding 1 int    | 967.83  | 96.78  |   1.61    |
|  Non colliding 1 int  | 983.01  |  98.3  |   6.84    |
|   Colliding 1 float   |  687.4  | 68.74  |   2.58    |
| Non colliding 1 float | 673.77  | 67.38  |   0.14    |
|    Colliding 2 int    |  825.5  | 82.55  |   0.51    |
|  Non colliding 2 int  |  825.5  | 82.55  |   0.82    |
|   Colliding 2 float   |  565.8  | 56.58  |   0.35    |
| Non colliding 2 float | 561.58  | 56.16  |   0.25    |

--------------------------------------------------
Group Mean: 76.13
Group Total: 6090.38
Group Standard Deviation: 15.46
Group Median: 75.64

**Comment by ScriptLineStudios** [comment]
RESULTS (OLD)

COLLISION: LINE-LINE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 589.82  | 58.98  |    0.6    |
| Non colliding | 600.94  | 60.09  |   1.38    |

--------------------------------------------------
Group Mean: 59.54
Group Total: 1190.76
Group Standard Deviation: 0.56
Group Median: 59.54



COLLISION: LINE-CIRCLE
==================================================
|     Name      |  Total  |  Mean  |  Std Dev  |
|---------------|---------|--------|-----------|
|   Colliding   | 591.59  | 59.16  |   1.73    |
| Non colliding | 694.22  | 69.42  |   0.53    |
| inside circle | 587.71  | 58.77  |   0.69    |

--------------------------------------------------
Group Mean: 62.45
Group Total: 1873.53
Group Standard Deviation: 4.93
Group Median: 59.16



COLLISION: LINE-POINT
==================================================
|         Name          |  Total  |  Mean  |  Std Dev  |
|-----------------------|---------|--------|-----------|
|    Colliding 1 int    | 1062.18 | 106.22 |   3.32    |
|  Non colliding 1 int  | 916.48  | 91.65  |   0.27    |
|   Colliding 1 float   | 656.09  | 65.61  |   0.78    |
| Non colliding 1 float | 662.65  | 66.26  |   0.43    |
|    Colliding 2 int    | 799.16  | 79.92  |   0.65    |
|  Non colliding 2 int  |  821.6  | 82.16  |   3.83    |
|   Colliding 2 float   | 560.78  | 56.08  |   5.12    |
| Non colliding 2 float | 546.54  | 54.65  |   0.28    |

--------------------------------------------------
Group Mean: 75.32
Group Total: 6025.47
Group Standard Deviation: 16.82
Group Median: 73.09

File: src_c/collisions.c [hunk 1]
```diff
@@ -83,15 +83,19 @@ pgIntersection_LineLine(pgLineBase *A, pgLineBase *B, double *X, double *Y,
...
+    double dx2 = line->x2 - Cx;
+    double dy2 = line->y2 - Cy;
+    double dx3 = line->x1 - line->x2;
+    double dy3 = line->y1 - line->y2;
+
+    double d1 = sqrt((dx * dx) + (dy * dy));
+    double d2 = sqrt((dx2 * dx2) + (dy2 * dy2));
+    double d3 = sqrt((dx3 * dx3) + (dy3 * dy3));
+
+    double width = 0.1;
```
Comment (Emc2356) [inline src_c/collisions.c:97]: should be 0.000001
0.1 is too big
my number is close to the max number of digits that a float can hold

**Review by Emc2356 — COMMENTED** [review]
just this one and i think it will be ok

File: src_c/collisions.c (hunk 1 above)
Comment (novialriptide) [inline src_c/collisions.c:97]: There appears to be whitespace at the end. Please remove.

Comment (novialriptide) [inline src_c/collisions.c:97]: Why is the width hardcoded? I think it'll be a great idea to allow the user to control a line's width.

File: src_c/collisions.c [hunk 2]
```diff
@@ -83,15 +83,19 @@ pgIntersection_LineLine(pgLineBase *A, pgLineBase *B, double *X, double *Y,
 static int
 pgCollision_LinePoint(pgLineBase *line, double Cx, double Cy)
 {
```
Comment (novialriptide) [inline src_c/collisions.c:85]: Instead of removing this equation entirely, shall we create an entirely new function with the its original old code?

**Review by novialriptide — CHANGES_REQUESTED** [review]
Thanks for the performance updates! I have a couple of questions though for our collaborators and you.

File: src_c/collisions.c (hunk 1 above)
Comment (itzpr3d4t0r) [reply src_c/collisions.c:97]: In theory we should do a visual test and see which number is the best in terms of precision and visual coherence

File: src_c/collisions.c [hunk 3]
```diff
@@ -83,15 +83,19 @@ pgIntersection_LineLine(pgLineBase *A, pgLineBase *B, double *X, double *Y,
...
+    double dy2 = line->y2 - Cy;
+    double dx3 = line->x1 - line->x2;
+    double dy3 = line->y1 - line->y2;
+
+    double d1 = sqrt((dx * dx) + (dy * dy));
+    double d2 = sqrt((dx2 * dx2) + (dy2 * dy2));
+    double d3 = sqrt((dx3 * dx3) + (dy3 * dy3));
+
+    double width = 0.000001;   
+    return d1+d2 >= d3-width && d1+d2 <= d3+width;
```
Comment (itzpr3d4t0r) [inline src_c/collisions.c:98]: You are using d1 and d2 only to sum them up and compare, and you are doing this twice. Do it once.
```suggestion
    double d = sqrt((dx * dx) + (dy * dy)) + sqrt((dx2 * dx2) + (dy2 * dy2));
    double d3 = sqrt((dx3 * dx3) + (dy3 * dy3));

    double width = 0.000001;   
    return d >= d3-width && d <= d3+width;
```

**Comment by itzpr3d4t0r** [comment]
I'm merging this as it's a good enough solution for now. We will need to think of a better solution in the future tho.

**Pull request merged**